### ISSUE TITLE
Why isn't u+263a being compiled?

### ISSUE THREAD

**Issue by averissimo** [googlei18n/noto-emoji#90]
![image](https://cloud.githubusercontent.com/assets/211358/22439101/4b582406-e726-11e6-815d-48f93dfa30f4.png)

**Comment by averissimo** [comment]
built it from 2dc2c4b (as well as released versions)

**Comment by averissimo** [comment]
and it seems to be compiled

![image](https://cloud.githubusercontent.com/assets/211358/22439469/a8d307d0-e727-11e6-8376-96e1e6bcae4a.png)

**Comment by dougfelt** [comment]
Works for me.

**Comment by dougfelt** [comment]
From the screenshot it looks like the character to the right of the missing character is U+263e.  Perhaps it is a different character that is not displaying?

**Comment by averissimo** [comment]
To reproduce the issue

_My system:_ Ubuntu 17.04 x64
_Chrome version:_ 57.0.2987.13 dev (64-bit)
_website:_ http://getemoji.com
_Font:__ latest release / master branch / or with the change I describe below

note: I'm using chrome as it maps correctly to the screen I showed before _(from an electron app)_

While building the font, I can see that the `third_party/color_emoji/emoji_builder.py` when processing u+263a it has the id `smileface`, instead of unixxxx as others.

Tried to backtrace that to fonttools `Lib/fontTools/agl.py` and removed the line:

`263A;smileface;WHITE SMILING FACE`

but that alone did no work.. I don't know anything about fonts, so I'd appreciate some help here :)

![image](https://cloud.githubusercontent.com/assets/211358/22448063/9290c076-e74e-11e6-9955-910895a08e45.png)

**Comment by dougfelt** [comment]
Hmmm.  @jungshik, care to weigh on on what Chrome is doing?

It looks like the sequence from the page you listed is U+263A U+FE0F, so this is trying to explicitly request an emoji representation, since in Unicode (2, 3, 4, 5beta) U+263A defaults to text presentation.  We do, in fact, have an entry for 263a fe0f in our variation selector cmap (at least in the font I built).  My guess is Chrome is still somehow not selecting our font but instead falling back to some other font.  Not sure why this would be the case.  Maybe our cmap is bad, or maybe Chrome just gets this wrong.

**Comment by DeeDeeG** [comment]
I'm not a googler or familiar with chrome or really that knowledgeable about fontconfig to be honest, (so I could be wrong! and I am curious what the answer would be from jungshik or someone who knows better than I do...) but it looks like a fontconfig issue to me.

Also seems like #36 which has an extensive discussion and some possible solutions.

I can give an in-depth explanation, from the best of my knowledge, if anyone is interested...

The page asks for Segoe UI Emoji (which we Linux users don't have installed) and so our systems look for any other font that can serve those glyphs. (Usually fontconfig is the subsystem that handles this -- fontconfig generally handles which fonts get used when on Linux).

(One can double-check that the font itself is fine by editing the page in the chrome dev-tools:)

### By default, the CSS on the page requests "Segoe UI Emoji" which I don't have, so I get 4 different fallback fonts)

![before](https://cloud.githubusercontent.com/assets/20157115/22454410/a463f084-e753-11e6-98b8-48fb199310a5.png)

### After (I edit the CSS to request "Noto Color Emoji" which is installed, and that's what I get)

![after](https://cloud.githubusercontent.com/assets/20157115/22454435/ce43337e-e753-11e6-9bfc-09a0f9cf52ca.png)

If averissimo repeats these steps and has the same result I have, I believe that helps confirm this is a fontconfig problem.

(Unless Chrome is somehow exacerbating the problem, *or* unless it is Chrome's responsibility to step in where fontconfig fails a user in this manner.) But I admit I am not an expert on this.

If so, we are basically asking fontconfig: "Hey fontconfig, give me *anything* that has a glyph at these codepoints" -- not specifically the user's shiny new color emoji font. 

Unfortunately Ubuntu has lots of fonts that contain glyphs at various (especially older) emoji codepoints, all in black and white outline style.

![liberationsanssmiley](https://cloud.githubusercontent.com/assets/20157115/22454931/58b7470e-e757-11e6-967f-53edfd241495.png)

And Ubuntu's developers haven't addressed this situation yet. Right now their fontconfig files do not mention any color emoji fonts, and so the color emoji fonts a user installs end up being "what's left" after the likes of Liberation Sans and Symbola have been exhausted, and as a consequence the user-installed fonts are usually searched last for glyphs.

This is a perennial issue for emoji fonts under Linux. I think it may be solvable by having a fontconfig configuration that makes NotoColorEmoji get used first over other fonts, but there might be minor issues in different apps. For averissimo's YakYak app, which seems to use CSS, maybe the app needs a change to more specifically ask for Noto Color Emoji? For a peek into the emoji-on-Linux rabbit hole, and some possible solutions, to usage in the browser, see #36.

**Comment by averissimo** [comment]
@DeeDeeG thanks for the very extensive response!! __I'm sorry I started doubting the build and font itself!__

Both chrome and in the initial example your solution works great! In chrome it is a problem of fontconfig (when it is installed)

In the example it was a matter of font priority in CSS.

I'm closing this issue as it solved my report and #36 still is open for any additional discussion.

Thanks to all!!1

ps. @z3ntu's comment in that thread show that lonely white emoji xD

**Comment by polarathene** [comment]
Ello, I've spent far too much time learning about fontconfig and a variety of emoji rendering problems, why they happen and how to solve them. I'll share this here for anyone else that gets almost there and ends up in this situation curious what's going on:

### Chrome has an extra layer of font management

This one was rightfully pointed out that it's a fallback issue, but it's not entirely fontconfigs fault though. Chrome preferences for fonts have an additional layer/override in the mix. If you've not touched these before they're "Custom" as the value for each generic font family as well as a "standard" font family. If you change from the "Custom" value, you lose that option and cannot reset it back afaik(not via the UI at least). I have no idea what the "Custom" value maps to(perhaps it is the preferred choice and forwards to fontconfig without having Chrome interfere?).

You can define each generic family to it's generic family(eg sans-serif to sans, serif to serif, fixed-width to monospace) instead of a specific font. This will query the same result that font-config would afaik...but it only seems to use the first result, like a `fc-match sans-serif`, not the remainder of the list for fallbacks, or querying for those glyphs/emoji(eg `fc-match sans-serif:charset=263A`). 

What happens instead, is if the font result initially retrieved lacks the glyph/emoji, it will fallback to the "Standard" font. Not great, since you lose the past information about sans-serif/serif/monospace family, you either provide a specific font or choose a generic family here as well, but it will always be that family. Not a big issue for emoji, but not great in other cases(eg a monospace falling back to sans-serif/serif for missing glyphs, such as asian language glyphs). Just keep in mind, that this little extra indirection can impact the font selection as a result. "Standard" font is also used when the website has no CSS defining a generic font fallback(there is a user-agent for some elements that will by default, but not everywhere), or when the font list/stack in a CSS declaration exhausts it's fonts for fallbacks(they don't fallback to the previous font-family rule). 

### Variation Selectors / Why do some fonts still not render their colour emoji?

One final gotcha that I have noticed, if the text content for an element begins with an emoji or regular text, that *may* affect font selection. I've seen this to be most notable with black/white(text) style vs the colour emoji style when a symbol/emoji has fonts that cover both styles. Not everyone is aware of it, but there's a thing called Variation Selectors, these have zero width so they're technically invisible and depending on implementation may be merged into the glyph/emoji as one character(eg a backspace deletes the whole character/emoji, otherwise only removes the VS, potentially changing appearance of the character/emoji). VS15 has a unicode codepoint of 0xFE0E to imply text presentation preference, and VS16 has a unicode codepoint of 0xFE0F to imply colour/emoji presentation preference. If neither are present, then it depends on what the default presentation is for that symbol/emoji(yes there is a variety with text representations as default, while others(newer) default to colour/emoji). Since font fallbacks don't seem to always respect this when applied, but might during search for the first fallback to satisfy the glyph/codepoint(emoji code/mapping), whichever font is matched first to provide that glyph will take priority and see no reason to find a fallback for the other form of presentation.

Firefox however has an internal emoji font setting in "about:config" which should be set to the bundled Twemoji font. This is a default fallback if no suitable fallback is provided afaik, but it also seems to respect FE0F/VS16 if the given font does not contain that support(some fonts can provide both variations). Firefox on Android [is even better than the linux desktop client](https://github.com/googlefonts/noto-emoji/issues/266#issuecomment-536254539) for some reason at respecting this. 

This was especially confusing [with this issue](https://github.com/googlefonts/noto-emoji/issues/265#issuecomment-539082348), where the bug was afaik related to how the developer(eosrei) placed that particular emoji and related ones into the HTML for the test site(The VS16 needs to be omitted from the skin tone combinations, so be careful of copy/paste!). 

Another good reference on this topic is my comments trying to help others with their VSCode rendering issues(complicated by Windows support). See [this one](https://github.com/be5invis/Sarasa-Gothic/issues/82#issuecomment-541302123) about ✔️ (`✔ vs ✔️`) and [this one](https://github.com/microsoft/vscode/issues/32840#issuecomment-541284810) about various other glyphs not rendering as expected.

### Getting 263A to render properly

If you adjust your font preferences for Chrome so that they're not "Custom", that should help. I'm not sure but it's possibly preferring Arial for the standard font at least. This is commonly mapped to Liberation Sans which is providing that problem emoji. So just like wanting to avoid DejaVu, you ideally would avoid any other fonts that contain black/white versions too(although it might be possible to filter/ignore these in a config via charset rule, I'm not sure).

By telling Chrome to use the generic font families, the requests will hit fontconfig as mentioned, and then the "Standard" font for fallback lookup for the emoji font you want to use. So using [the usual config](https://github.com/googlefonts/noto-fonts/issues/1212#issuecomment-414405453) or [commonly cited one](https://github.com/googlefonts/noto-emoji/issues/36#issuecomment-343655365)(possibly with strong binding instead of weak) should work fine in most cases.

You can also blacklist fonts like DejaVu families:

```XML
 <selectfont>
        <rejectfont>
            <pattern>
                <patelt name="family">
                    <string>DejaVu Sans</string>
                </patelt>
            </pattern>
            <pattern>
                <patelt name="family">
                    <string>DejaVu Serif</string>
                </patelt>
            </pattern>
            <pattern>
                <patelt name="family">
                    <string>DejaVu Sans Mono</string>
                </patelt>
            </pattern>
        </rejectfont>
    </selectfont>
```

Do note that, that Chrome should be restarted to ensure it takes affect correctly. Elsewhere on the system may require a reboot/relogin, there also seems to be other fonts overriding selection on the system for me(like window titlebars), that needs the blacklist/rejection snippet above among some other changes afaik, still experimenting with config there.

### Tips - Identify font used per glyph in browser & Inspect font fallback list generated via configs

If you weren't aware, you can verify the 263A font via browser dev tools. Firefox is a bit more useful since rolling over the listed fonts used for rendering will highlight the glyphs related to that font. For both, accuracy of listed fonts may depend on whatever HTML element is selected in the HTML inspector panel.

If you'd like to a list of fonts that are generated from your fontconfig for a pattern/query, you can run the following command:

`FC_DEBUG=4 fc-match monospace | grep -A2 "done"`

### Some gotchas with matching behaviour:

After that, afaik, fonts after the matched one (`fc-match monospace` in this case) should exist in the lists you'd get with `--sort` option, but only if they provide coverage that prior fonts did not(strong binding or other rules might affect that), the rest get omitted, then after that the rest is not from fc-match list that was generated but going through the fonts on the system I guess. I explain that [here](https://bbs.archlinux.org/viewtopic.php?pid=1867652#p1867652) on an Arch forum thread. 

One other thing to keep in mind is that you likely have a `/etc/fonts/conf.d/49-sansserif.conf` and this file will assign a requested font(if not a generic family) sans-serif if it did not have a family assigned earlier. Probably doesn't matter much for emoji, but it does mean font families can end up generating the wrong fallbacks, quite common for monospace/programming fonts. On Arch, currently, the Noto family(serif/sans-serif/mono) have font configs in the 6x range, which is too late for this assignment, so you end up with your monospace and serif having fallbacks triggered for sans-serif before monospace is added into the list(if you use the debug command shown above, you'll probably see some results repeated along the way due to how configs/rules have been defined).

It also happened for [this user](https://bbs.archlinux.org/viewtopic.php?pid=1867659#p1867659) where their preferred fonts were not being respected in their config. That's possibly because of other configs misbehaving and not following the proper usage of those numbers, assuming that they're just to be treated as a priority without any other purpose. [They do have logical purpose](http://www.linuxfromscratch.org/blfs/view/svn/x/tuning-fontconfig.html) however. I share a few ways to work around that issue there and prioritize the desired fonts.
